### PR TITLE
Added support to abort test early if a threshold fails

### LINKED ISSUE #449: Add option to thresholds to abort test early if it evaluates to false midway
Thresholds are today evaluated continuously throughout a test, every 2s to be precise [1].

Every cycle we loop through the defined thresholds and check whether the threshold JS expression (yes, the `"p(95)<100"` etc. is JS code so you could get creative with your thresholds 😄) evaluates to true or false given the current collected samples and the result of this evaluation is stored in the metric's `Tainted` property [2].

If you're running k6 with `-o json=output.json` to output results to a JSON file you'll see a `tainted` property for metric samples being emitted regularly, and the value can fluctuate between true and false throughout the test.

Internally though, k6 will only care about the last evaluation [3]. The result of which is what determines if a threshold passed or failed the test.

There's been some discussion to implement something similar to the old [`--abort-on-taint`](https://github.com/loadimpact/k6/commit/9ac1ff38a4c0c5749880eba9639deee1df2fb681) flag. The most promising I think is to extend the threshold definition to allow for an `abortOnTaint` option per threshold expression, extending the definition from:

```js
export let options = {
    thresholds: {
        http_req_duration: ["p(95)"]
    }
};
```

to also allow:

```js
export let options = {
    thresholds: {
        http_req_duration: [{ threshold: "p(95)<100", abortOnTaint: true }]
    }
};
```

This will not break backwards compatibility and also work well for the three different execution modes we want to support; local, cloud and clustered.

Things to consider:
- Should there be a grace period in the beginning of the test until we have "enough" samples to run a meaningful threshold evaluation?

Thoughts and comments welcome.

[1] - https://github.com/loadimpact/k6/blob/master/core/engine.go#L41
[2] - https://github.com/loadimpact/k6/blob/master/stats/stats.go#L176
[3] - https://github.com/loadimpact/k6/blob/master/core/engine.go#L276

### PR DESCRIPTION
**Pull request by antekresic**
Closes #449 

Did not add mentioned grace period (check issue), should be trivial to add if we decide to do it.

### CODE REVIEW

File: stats/thresholds.go [hunk 1]
```diff
@@ -81,26 +83,62 @@ func (t *Threshold) Run() (bool, error) {
 	return b, err
 }
 
+type ThresholdConfig struct {
+	Threshold    string `json:"threshold"`
+	AbortOnTaint bool   `json:"abortOnTaint"`
```
Comment (robingustafsson) [inline stats/thresholds.go]: When reading the code now I think `abortOnFail`, that you've used in other places, sounds better than `abortOnTaint` (and will probably be easier to understand on first look for people reading script code), so I'm of the opinion that we should change the user facing parameter to be called `abortOnFail`.

Comment (antekresic) [reply stats/thresholds.go]: Totally agree, changed all `abortOnTaint` references to `abortOnFail` for consistency.

**Comment by antekresic** [comment]
Hey @robingustafsson,
I added the grace functionality as well to this PR.

**Comment by robingustafsson** [comment]
@antekresic Great, thanks! I'll have a look and give it a try.

**Comment by robingustafsson** [comment]
LGTM, merging!

**Pull request merged**